### DISCUSSION TITLE
Adding a custom tab for a Device

### DISCUSSION THREAD

**Discussion by FilipKatulski** [netbox-community/netbox#11884] [Help Wanted!] [answered]
Dear Community! 
I am a fairly new Netbox and Django user. I apologize if the question is too trivial for this forum. 

I would like to add a custom tab for every Device, next to the "Changelog" tab, which would display the output of the "ipmitool". I cannot target the proper files that are responsible for displaying the "Changelog" and "Journal" tabs. 

Moreover, I have tried to create a workaround solution - adding an additional Custom button next to the "Add Components" and "Edit" buttons, within the 'netbox/templates/dcim/device/base.html', however the button does not appear after rebuilding the Netbox. The User account has the proper perms to display that custom button.

1. Where can I modify Tabs sections for a Device? 
2. Is there an additional setting/configuration for buttons in the 'base.html' file which have to be modified?

Thanks!

**Comment by candlerb** [comment, marked as answer]
> I cannot target the proper files that are responsible for displaying the "Changelog" and "Journal" tabs.

You must not modify Netbox core: if you do, you are forking the project and you are 100% responsible for maintaining your fork going forwards.  You will find it *very* difficult, if not impossible, to merge in future developments in Netbox.  Large refactorings between versions are commonplace.

What you should do instead is to write a [plugin](https://docs.netbox.dev/en/stable/plugins/).  This has a clearly-defined set of things which you are allowed to do, including being able to [add tabs](https://docs.netbox.dev/en/stable/plugins/development/views/#additional-tabs) to existing model views.  Whilst there are no guarantees that you won't have to make any changes to your plugin in future, it will at least have a clearly-defined set of touchpoints, and anything which affects you will be called out in the release notes.

For custom buttons, you should use [custom links](https://docs.netbox.dev/en/stable/plugins/) for the same reason. Here is a Custom Link definition I made for IPMI access:

```
# Link text
{% if object.interfaces.filter(mgmt_only=True).count() == 1 and object.interfaces.get(mgmt_only=True).ip_addresses.count() == 1 %}Open IPMI{% endif %}

# Link URL
https://{{ object.interfaces.get(mgmt_only=True).ip_addresses.first().address.ip }}/
```

(Note: hasn't been tested for a while)

To reiterate: never, never touch base.html or any existing source files within Netbox!

**Reply by FilipKatulski** [reply]
Thank you for these suggestions! I will try to use them.
